### PR TITLE
chrome_extensions: Refactor the table, add tests

### PR DESCRIPTION
**Pull request by alessandrogario**
This is a rewrite of the `chrome_extensions` and `chrome_extensions_content_scripts` virtual table implementations that makes them testable, adds tests, fixes problems with missed data in the results, and expands the schema. 

The schema additions for `chrome_extensions_content_scripts`: 
- include columns for `browser_type` (because it supports several Chrome-derived browsers)
- the Chrome profile path
- the path to the extension folder
- a boolean value for whether the extension is referenced from the profile's Preferences file. 

The schema additions for `chrome_extensions`: 
- `browser_type`
- `profile_path`
- the `locale` column is replaced by `default_locale` and `current_locale` 
- many columns sourced from the manifest.json
- a hash of the manifest file itself in order to monitor it for changes. 

These tables replace the `opera_extensions` tables, but in doing so, add support for querying Opera extensions on Windows.

~**Note** Build requirement on macOS is the 10.14 or higher SDK:
`cd build; cmake -DCMAKE_OSX_DEPLOYMENT_TARGET=10.14 -DOSQUERY_BUILD_TESTS=ON ..`~ Removed `std::optional` to avoid this requirement.

Fixes: #6069

### CODE REVIEW

File: specs/chrome_extensions.table
Comment (theopolis) [inline specs/chrome_extensions.table:12]:
```suggestion
    Column("default_locale", TEXT, "Default locale supported by extension", aliases=["locale"]),
```
and remove the `locale` column below.

**Review by theopolis — CHANGES_REQUESTED** [review]
Look though all of the potential status checks and TODOs.

File: osquery/filesystem/filesystem.cpp
Comment (theopolis) [inline osquery/filesystem/filesystem.cpp:110]: What's the goal with this new file reading method? Is there a limitation with the existing methods?

Comment (theopolis) [reply osquery/filesystem/filesystem.cpp:110]: Heads up, still interested in this question @alessandrogario :D

**Comment by mike-myers-tob** [comment]
Also closes https://github.com/osquery/osquery/issues/6069

File: osquery/filesystem/filesystem.cpp
Comment (alessandrogario) [reply osquery/filesystem/filesystem.cpp:110]: > Heads up, still interested in this question @alessandrogario :D

The current readFile() methods that we have under filesystem.cpp are pretty big and rather complex; an additional issue was that they don't return many details when something fails - just a text string.

This was an attempt to use a simpler function, with better error handling and the more modern `Expectd<ValueType, ErrorType>` way of handling returned values that was implemented in the experimental branch

Comment (zwass) [reply osquery/filesystem/filesystem.cpp:110]: Do you anticipate refactoring to eliminate the old method in the codebase? I'd like to avoid the proliferation of different functions for achieving the same goal.

**Review by zwass — COMMENTED** [review]
I'd like to see @theopolis concern resolved before we merge this.

File: osquery/filesystem/filesystem.cpp
Comment (alessandrogario) [reply osquery/filesystem/filesystem.cpp:110]: > Do you anticipate refactoring to eliminate the old method in the codebase? I'd like to avoid the proliferation of different functions for achieving the same goal.

I think it would be useful to refactor the filesystem functions so that they become smaller and easier to work with; something similar to the fs interface we have in BPF (`osquery/events/linux/bpf/ifilesystem.h`) could prove really useful when implementing tests, because it lets us swap the implementation with a mocked class. An example for this usage can be also found in the BPF code: `osquery/events/tests/linux/bpf/mockedfilesystem.cpp`

File: osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp
Comment (directionless) [inline osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: Should this be in the if block?

File: specs/chrome_extensions.table
Comment (directionless) [inline specs/chrome_extensions.table:26]: Is there value in preserving this? (I think most of the windows tables just convert, and return the converted value)

File: osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp
Comment (alessandrogario) [reply osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: The intent here was to always initialize the column; in case the version field is not present in the properties, it will initialize the version column to an empty string

File: specs/chrome_extensions.table
Comment (alessandrogario) [reply specs/chrome_extensions.table:26]: The Webkit format is pretty strange, so it's not immediate to convert it to/from unix timestamps. The reason I kept it there is because it lets users look for timestamps they may have seen in their local files

File: osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp
Comment (directionless) [reply osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: So this has the effect of setting it to an empty string, instead of leaving it null? I'm not sure I have a strong feeling either way. Initially, null seems more correct.

I guess I'd have ended up with:
```
      if (version_it == extension.properties.end()) {
        LOG(ERROR)
            << "The version property for the following extension is missing: "
            << extension.path;
        row["version"] = ""
      } else {
        row["version"] = SQL_TEXT(version_it->second);
      }
```

Comment (alessandrogario) [reply osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: In the past we had some crashes due to code assuming that columns were present when in fact they were being left missing by the implementation; this gets harder when the implementation differs based on platform. Since then, I think we have all been trying to make sure that table code always populate all the columns listed in the schema

Comment (directionless) [reply osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: I remember issues based on whether or not a column is defined, platform specific columns end up being very hard to write generalized queries for. (There's a hack!)

But I don't remember crashes about null values. 

I feel like sqlite should support nulls cleanly. Else it's impossible to differentiate between an empty string and no data.

Comment (alessandrogario) [reply osquery/tables/applications/chrome/chrome_extension_content_scripts.cpp:42]: The crashes were caused by code inside osquery running queries against other tables, then trying to access columns that were not being populated by one or the other implementation.

sqlite does support NULL column values, but osquery was initially designed to always convert everything to string, causing that feature (and real types) to get lost.

**Comment by mike-myers-tob** [comment]
Did the `manifest.json` file move since this PR was started? osquery seems to look in many locations where the file does not exist and then it generates errors.

```
osquery> select uid, identifier from chrome_extension_content_scripts;
W0226 17:15:44.200911  1620 virtual_table.cpp:961] The chrome_extension_content_scripts table returns data based on the current user by default, consider JOINing against the users table
W0226 17:15:44.200911  1620 virtual_table.cpp:976] Please see the table documentation: https://osquery.io/schema/#chrome_extension_content_scripts
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\web_store\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\feedback\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\cryptotoken\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\cloud_print\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\pdf\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\network_speech_synthesis\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
E0226 17:15:44.248322  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files\Google\Chrome\Application\88.0.4324.190\resources\hangout_services\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Google\Chrome\User Data\Default
```

Also I had MS Edge on this VM but it had been updated a few times since it was originally installed. This appeared to cause problems too, as osquery still looked for manifest.json for each of the old versions:

```
E0226 17:15:44.263523  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\edge_clipboard\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.263523  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\media_internals_services\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.263523  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\edge_collections\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\microsoft_web_store\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\edge_feedback\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\microsoft_voices\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\cryptotoken\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\edge_pdf\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\87.0.664.55\resources\webrtc_internals\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
E0226 17:15:44.279049  1620 utils.cpp:535] Failed to read the following manifest.json file: C:\Program Files (x86)\Microsoft\Edge\Application\85.0.564.67\resources\hangout_services\manifest.json. The extension was referenced by the following profile: C:\Users\mmyers\AppData\Local\Microsoft\Edge\User Data\Default
```

(Edge was on `Version 88.0.705.81` when this query was made).

**Comment by directionless** [comment]
What's the state on this? I'm currently holding 4.7.0 for it

**Comment by mike-myers-tob** [comment]
> What's the state on this? I'm currently holding 4.7.0 for it

Alessandro addressed the last requested change from feedback, so, we need an approval.

**Review by theopolis — APPROVED** [review]

**Pull request merged**
